flocker/node/test/test_deploy.py

```diff
@@ -2291,13 +2336,371 @@ def test_dataset_id_generated(self):
         d = api.calculate_necessary_state_changes(desired, actual, u"node")
         result = self.successResultOf(d)
         # CreateVolume:
-        dataset = result.changes[0].changes[0].volume.dataset
+        dataset = result.changes[0].changes[0].dataset
         # StartApplication:
         dataset2 = result.changes[1].changes[0].application.volume.dataset
         # New UUID was generated, but only once:
         self.assertEqual(UUID(dataset.dataset_id), UUID(dataset2.dataset_id))
 
 
+class DeployerCalculateNecessaryStateChangesDatasetOnlyTests(
+        SynchronousTestCase):
+    """
+    Tests for ``Deployer.calculate_necessary_state_changes`` when only
+    datasets are involved which are not attached to applications.
+    """
+    def test_dataset_created(self):
+        """
+        ``Deployer.calculate_necessary_state_changes`` specifies that a new
+        dataset must be created if the desired configuration specifies
+        that a dataset that previously existed nowhere is going to be
+        on this node.
+        """
+        hostname = u"node1.example.com"
+
+        current = Deployment(nodes=frozenset({
+            Node(hostname=hostname, applications=frozenset()),
+        }))
+
+        api = Deployer(
+            create_volume_service(self),
+            docker_client=FakeDockerClient(units={}),
+            network=make_memory_network()
+        )
+
+        node = Node(
+            hostname=hostname,
+            other_manifestations=frozenset({MANIFESTATION}),
+        )
+        desired = Deployment(nodes=frozenset({node}))
+
+        calculating = api.calculate_necessary_state_changes(
+            desired_state=desired,
+            current_cluster_state=current,
+            hostname=hostname,
+        )
+
+        changes = self.successResultOf(calculating)
+        expected = Sequentially(changes=[
+            InParallel(changes=[CreateDataset(
+                dataset=MANIFESTATION.dataset)])])
+        self.assertEqual(expected, changes)
+
+    def test_volume_wait(self):
+        """
+        ``Deployer.calculate_necessary_state_changes`` specifies that the
+        dataset previously stored on another node must be waited for, in
+        anticipation of that node handing it off to us.
+        """
+        # The application is not running here - therefore there is no container
```

Comment:
If it's in `other_manifestations` on `node2` then there's no application involved at all, right?  Otherwise it would be attached to the application instead.